spec/unit/module_loader/puppetfile_spec.rb

```diff
@@ -0,0 +1,297 @@
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', File.join(basedir, 'vendor'), module_opts, anything).and_call_original
+
+      expect { subject.add_module('puppet/test_module', module_opts) }.to change { subject.modules }
+      expect(subject.modules.collect(&:name)).to include('test_module')
+    end
+
+    it 'should accept non-Forge modules with a valid absolute :install_path option' do
+      install_path = File.join(basedir, 'vendor')
+
+      module_opts = {
+        install_path: install_path,
+        git: 'git@example.com:puppet/test_module.git',
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', install_path, module_opts, anything).and_call_original
+
+      expect { subject.add_module('puppet/test_module', module_opts) }.to change { subject.modules }
+      expect(subject.modules.collect(&:name)).to include('test_module')
+    end
+
+    it 'should reject non-Forge modules with an invalid relative :install_path option' do
+      module_opts = {
+        install_path: '../../vendor',
+        git: 'git@example.com:puppet/test_module.git',
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', File.join(basedir, 'vendor'), module_opts, anything).and_call_original
+
+      expect { subject.add_module('puppet/test_module', module_opts) }.to raise_error(R10K::Error, /cannot manage content.*is not within/i).and not_change { subject.modules }
+    end
+
+    it 'should reject non-Forge modules with an invalid absolute :install_path option' do
+      module_opts = {
+        install_path: '/tmp/mydata/vendor',
+        git: 'git@example.com:puppet/test_module.git',
+      }
+
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', File.join(basedir, 'vendor'), module_opts, anything).and_call_original
+
+      expect { subject.add_module('puppet/test_module', module_opts) }.to raise_error(R10K::Error, /cannot manage content.*is not within/i).and not_change { subject.modules }
+    end
+
+    it 'should disable and not add modules that conflict with the environment' do
+      env = instance_double('R10K::Environment::Base')
+      mod = instance_double('R10K::Module::Base', name: 'conflict', origin: :puppetfile)
+      allow(mod).to receive(:origin=).and_return(nil)
+      loader = described_class.new(basedir: basedir, environment: env)
+      allow(env).to receive(:'module_conflicts?').with(mod).and_return(true)
+
+      allow(R10K::Module).to receive(:new).with('test', anything, anything, anything).and_return(mod)
+      expect { loader.add_module('test', {}) }.not_to change { loader.modules }
+    end
+  end
+
+  describe '#purge_exclusions' do
+    let(:managed_dirs) { ['dir1', 'dir2'] }
+    subject { R10K::ModuleLoader::Puppetfile.new(basedir: '/test/basedir') }
+
+    it 'includes managed_directories' do
+      expect(subject.send(:determine_purge_exclusions, managed_dirs)).to match_array(managed_dirs)
```

Comment:
Is the reason you need to use `send` because `determine_purge_exclusions` is private? 

I feel like it might make more sense to test the behavior we want to see rather than reaching into private methods? Like, as is, this test doesn't seem very valuable to me..

Reply:
Did I not do this?

Reply:
Looks like no. I believe after I push my local changes up, this will be the one outstanding change that I still need to make, however I won't get to it till next week.

Reply:
Is this still something you're planning to do as part of this PR?

Reply:
Fwiw, as much as I agree that this test isn't super useful, I also think it would be good to get this in, so if this is the only pending comment, I'm inclined to leave writing new tests for another time. We'll get better validation running it in real environments.

Reply:
Do we want/need a ticket for writing better automated tests?

The only other pending comments were around code comments that I found confusing, one of them being straight up not true (the one about the `puppetfile_content` method). so, not the biggest deal, but could be extra confusing for future us. ¯\_(ツ)_/¯